### DISCUSSION TITLE
Require use of shipping calculator before "proceed to checkout"?

### DISCUSSION THREAD

**Discussion by valtandor** [woocommerce/woocommerce#53615] [Checkout Flow]
I'm the webmaster at a small business, and we use the "Custom shipping quote" plug-in to give customers the option of requesting a quote. This is intended to cover a variety of special situations, such as large orders, Saturday deliveries, courier delivery to Alaska/Hawaii/US territories/etc, and overseas deliveries of all kinds. We also allow the "Local pickup" option for people who intend to be passing by our physical store.

For small orders to be sent by ordinary means within the US, we use Weight Based Shipping, with various pre-priced delivery options configured in that based on delivery zone and package weight. Of course, before a standard price can be offered to the customer, the customer has to enter a delivery area (country, state/region, city, postcode or ZIP code) and then the shipping calculator does its thing.

Here's my problem. Because Local Pickup is an option, I can't use the, "Hide shipping costs until an address is entered," option (WooCommerce -> Settings -> Shipping -> Shipping Settings). Moreover, "Custom shipping quote" is the default option until a delivery address is chosen.

To guide the customer, here is what I would like to do:

1. Disable or suppress the, "Proceed to checkout," button unless _either_ a delivery area is specified (by entering the information into, and running, the shipping calculator) , _or_ a Local Pickup location is chosen
2. Make it so that no shipping option is selected by default - the customer must always take some action, but, "Proceed to checkout," should not be enabled unless a shipping option is selected. At least, the web site administrator should be able to do this as an option in the Wordpress GUI.
3. Don't show the, "Custom shipping quote," shipping option unless a delivery area is specified (by entering the information into, and running, the shipping calculator), in the same way that the options offered by Weight Based Shipping aren't shown unless a delivery area is so specified

I gather that item 2 can be done through a custom functions.php, or through custom PHP or CSS code put into the WordPress editor, but items 1 and 3 are much harder, especially item 1, for which the appearance and behaviour of the, "Proceed to checkout," button appears to be locked except for such cosmetic aspects as the precise text of that button. Item 3 is perhaps a matter for the developer of the Custom Shipping Quote plugin/extension.

But before I lose too much more sleep over this, have these things been considered before, and rejected, perhaps after a trial to see how well they work for businesses?

**Comment by opr** [comment]
Hi @valtandor thanks for posting and proving a detailed use case!

This isn't something we've considered before, actually. It's an interesting use-case and it sounds like something we could support, however I don't think it's something we'd make a GUI control for it, it's likely that it'd use custom code, unless we find widespread desire for this.

> Here's my problem. Because Local Pickup is an option, I can't use the, "Hide shipping costs until an address is entered," option (WooCommerce -> Settings -> Shipping -> Shipping Settings). Moreover, "Custom shipping quote" is the default option until a delivery address is chosen.

We are changing the way this works in https://github.com/woocommerce/woocommerce/pull/53466, so only local pickup options show until an address is entered.

> Disable or suppress the, "Proceed to checkout," button unless either a delivery area is specified (by entering the information into, and running, the shipping calculator) , or a Local Pickup location is chosen

Unfortunately there's no programmatic way to disable the button at the moment, you will have to revert to CSS/DOM manipulation until we implement something. I added https://github.com/woocommerce/woocommerce/issues/53751 to track this.

> Make it so that no shipping option is selected by default - the customer must always take some action, but, "Proceed to checkout," should not be enabled unless a shipping option is selected. At least, the web site administrator should be able to do this as an option in the Wordpress GUI.

As mentioned, this is not something we would implement a GUI for, however as you also noted you can achieve this with the existing hooks.

> Don't show the, "Custom shipping quote," shipping option unless a delivery area is specified (by entering the information into, and running, the shipping calculator), in the same way that the options offered by Weight Based Shipping aren't shown unless a delivery area is so specified

I believe this will be solved by https://github.com/woocommerce/woocommerce/pull/53466 :)

**Comment by pierorocca** [comment]
Hi @valtandor thanks for bringing this use case forward. I want to get more insight into:

> we use the "Custom shipping quote" plug-in to give customers the option of requesting a quote. 

Could you elaborate on this workflow and where you present this option to customers? Shipping is a really complex, interesting area that's often overlooked so I'd like to learn more. Thanks for the education.

**Comment by valtandor** [comment]
@pierorocca and @opr

Thanks for your patience!

@pierorocca

> Could you elaborate on this workflow and where you present this option to customers?

We present "Custom shipping quote" on the Cart page. I think it's also available on the Checkout page, but my preference is for it to be meaningfully available on the Cart page. In my business, **I believe strongly in a, "No nasty surprises for the customer," policy**: so strongly, that although we're in Texas where the norm is to quote prices excluding sales tax, we include sales tax in our over-the-counter prices. (We can't do this for online, because tax depends on what state the goods are being shipped to.)

**The reason I want "Local pickup" to be available to the customer even when no address is entered is for customer privacy.** Customers shouldn't have to disclose more information about themselves than is necessary for a successful transaction. We live in an age where privacy is important to people, and they get inundated with unsolicited mail. If a customer wants to pick up, I don't need to know a shipping address, and I don't want to be responsible for securely storing any of a customer's personal information that I don't have a proper reason to collect in the first place. (Billing address may be needed for card verification, but that could be a PO Box or other non-street address.)

@opr

Do you know if #53466 has made it into production yet? It looks like it's made it into _trunk_, but perhaps it hasn't been released. What release number am I looking for? It says "Milestone: 9.8.0" on the issue if I understand correctly.

I would like to try the results of #53466 to see whether it suppresses "Custom shipping quote" as well as other options if a destination area (postcode, etc.) hasn't been selected by the customer.

You can appreciate that for production purposes I would prefer not to deploy a bleeding-edge (development) version of the WooCommerce plugin into my web store!

**Reply by opr** [reply]
> @opr
>
> Do you know if https://github.com/woocommerce/woocommerce/pull/53466 has made it into production yet? It looks like it's made it into trunk, but perhaps it hasn't been released. What release number am I looking for? It says "Milestone: 9.8.0" on the issue if I understand correctly.
>
> I would like to try the results of https://github.com/woocommerce/woocommerce/pull/53466 to see whether it suppresses "Custom shipping quote" as well as other options if a destination area (postcode, etc.) hasn't been selected by the customer.

It should be in release 9.8 yes, in the meantime perhaps you could install the [WC Beta Tester](https://github.com/woocommerce/woocommerce/releases/tag/wc-beta-tester-3.0.0) plugin (download the ZIP from this release page and install it as a regular plugin). This comes with the caveat that this is not supported and should only be used in a testing environment.

Once this is installed and activated go to WooCommerce -> Live Branches and install and activate the `trunk` branch.

Let me know if you have any feedback on that PR - you can leave the feedback directly on the PR itself. 

Thanks!

**Comment by pierorocca** [comment]
Thanks @valtandor for sharing your use cases and the principles behind them. That's a noble position.

#53466 appears to address the majority of your problems. 

The option to disable "proceed to checkout" based on conditions on the cart is more tricky. Despite privacy issues, customers also overwhelmingly want easy options and to use and re-use their preferred stored addresses and payment methods like Apple Pay and Google Pay and many others across the globe. What I think could help in your desired flow is a multi-step checkout that allows progression based on meeting the required conditions before "proceeding to payment". Think that over and let me know if a multi-step option would suit your site better than the one-page model.

Thanks again.

**Reply by valtandor** [reply]
@pierorocca

Thank you for the kind words! To be clear, I have no problem if a customer wants to register on the web site (we have a user account option) and save one or more addresses for billing, delivery, or both. But so far, most of our customers have not availed themselves of that option. So the issue is not with a customer choosing to save an address; the issue is making sure the customer isn't _required_ to provide a delivery address (or even a delivery area) to proceed past the Cart page and onto the Checkout page if local pickup is what's chosen.

**Reply by pierorocca** [reply]
Gotcha. It does appear #53466 will help with that and agree the Cart shouldn't force anything to progress to Checkout...except to have at least 1 item in the cart. :)
